sql/rowexec: TestUncertaintyErrorIsReturned failed

[(sql/rowexec).TestUncertaintyErrorIsReturned failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2188154&tab=buildLog) on [master@7ad048f9935dd565dede9153ebf441e4346e004f](https://github.com/cockroachdb/cockroach/commits/7ad048f9935dd565dede9153ebf441e4346e004f):

```
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8

Goroutine 528 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1188 +0x86
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.StartableFn.Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:47 +0x5b
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:321 +0x330
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:346 +0x98
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Start()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:109 +0x26a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:130 +0x431
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x150
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:127 +0xcb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:574 +0x4bc
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:630 +0xa1
  github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbd8
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:211 +0x135
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:61 +0xd9
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:55 +0x80
  google.golang.org/grpc.chainUnaryServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0x110
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:632 +0x1d8
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8
==================
FAIL	github.com/cockroachdb/cockroach/pkg/sql/rowexec	18.673s

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestUncertaintyErrorIsReturned PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestUncertaintyErrorIsReturned.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/rowexec).TestUncertaintyErrorIsReturned failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2191811&tab=buildLog) on [master@22a31b641abbb5ec59023eb73dde63146159b515](https://github.com/cockroachdb/cockroach/commits/22a31b641abbb5ec59023eb73dde63146159b515):

```
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.StartableFn.Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:47 +0x5b
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:321 +0x330
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:346 +0x98
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Start()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:109 +0x26a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:130 +0x431
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x150
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:127 +0xcb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:574 +0x4bc
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:630 +0xa1
  github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbd8
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:211 +0x135
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:61 +0xd9
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:55 +0x80
  google.golang.org/grpc.chainUnaryServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0x110
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:632 +0x1d8
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8
==================
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestUncertaintyErrorIsReturned463545211
        --- FAIL: TestUncertaintyErrorIsReturned/vectorize=true/SELECT_*_FROM_t_JOIN_onerow_ON_t.x_=_onerow.x (0.09s)
            testing.go:853: race detected during execution of test
        testing.go:853: race detected during execution of test
    testing.go:853: race detected during execution of test
=== RUN   TestUncertaintyErrorIsReturned/vectorize=true
    --- FAIL: TestUncertaintyErrorIsReturned/vectorize=true (0.34s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestUncertaintyErrorIsReturned PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestUncertaintyErrorIsReturned.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/parallel_unordered_synchronizer.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer.go
@@ -23,9 +23,10 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
-// unorderedSynchronizerMsg is a light wrapper over a coldata.Batch sent over a
-// channel so that the main goroutine can know which input this message
-// originated from.
+// unorderedSynchronizerMsg is a light wrapper over a coldata.Batch or metadata
+// sent over a channel so that the main goroutine can know which input this
+// message originated from.
+// Note that either a batch or metadata must be sent, but not both.
 type unorderedSynchronizerMsg struct {
 	inputIdx int
 	b        coldata.Batch
@@ -237,6 +238,11 @@ func (s *ParallelUnorderedSynchronizer) init(ctx context.Context) {
 					// In case of a zero-length batch, proceed to drain the input.
 					fallthrough
 				case parallelUnorderedSynchronizerStateDraining:
+					// Create a new message for metadata. The previous message cannot be
+					// overwritten since it might still be in the channel.
+					msg = &unorderedSynchronizerMsg{
+						inputIdx: inputIdx,
+					}
 					if input.MetadataSources != nil {
 						msg.meta = input.MetadataSources.DrainMeta(ctx)
 					}

--- a/pkg/sql/rowexec/processors_test.go
+++ b/pkg/sql/rowexec/processors_test.go
@@ -852,7 +852,6 @@ func TestUncertaintyErrorIsReturned(t *testing.T) {
 	testutils.RunTrueAndFalse(t, "vectorize", func(t *testing.T, vectorize bool) {
 		vectorizeOpt := "off"
 		if vectorize {
-			skip.WithIssue(t, 52948)
 			vectorizeOpt = "on"
 		}
 		for _, testCase := range testCases {

